### ISSUE TITLE
A whole table will be deleted if primary key is 0 or not set when using db.Delete()

### ISSUE THREAD

**Issue by cjun714** [jinzhu/gorm#2025]
It has been reported since #1152, and document also emphasis:

> GORM will use the primary key to delete the record, if primary field’s blank, GORM will delete all records for the model.

I am really shocked when I first read #1152 occasionally.
I can't understand why this implementation was done, it's too dangerous and quit counter-intuition:
- Nobody will think delete(Pk=0) will delete a whole table.
- Lots of people will never read whole document, especially when they thought they already "knew that case".
- People always refer previous knowledge instead of reading document.
- People always refer other people's code snippet/introduction slides/blog to do their task instead of checking manual unless they meet issue or trouble.

Above is quit common in a real world.

Below code is quit common, but so dangerous in Gorm.
``` go
id := getIntParaFromWebRequest("id")  // implemented by web engineer, default return 0
e := dao.DeleteUser(id)
---------------------------------------------------
func DeleteUser(id uint) error { // implemented by dao engineer who hasn't read manual
	user := User{ID: id}
	return db.Delete(&user).Error
}
```
 A nuclear  button should never be put on your dinner table with a warning message posted somewhere.

### What version of Go are you using (`go version`)?
Latest

### Which database and its version are you using?
Sqlite3

### Please provide a complete runnable program to reproduce your issue. **IMPORTANT**

```go
package main

import (
	"github.com/jinzhu/gorm"
	_ "github.com/jinzhu/gorm/dialects/sqlite"
)

var db *gorm.DB

func init() {
	var err error
	db, err = gorm.Open("sqlite3", "test.db")
	if err != nil {
		panic(err)
	}
	db.LogMode(true)
	db.AutoMigrate(&User{})
}

func main() {
	db.Create(&User{Name: "May"})
	db.Create(&User{Name: "Joe"})
	db.Delete(&User{ID: 0}) // cause deleting all records in user table

	db.Create(&User{Name: "May"})
	db.Create(&User{Name: "Joe"})
	db.Delete(&User{Name: "Joe"}) // cause deleting all records in user table
}

type User struct {
	ID   uint
	Name string
}
```

**Comment by mollrow** [comment]
Yeah this seems very bad. I'm still pretty new to Go and make mistakes often so having this be a special case like `db.DeleteAll` would be good.

**Comment by pedromorgan** [comment]
Caught me out as well.

**Comment by emirb** [comment]
Duplicate of https://github.com/jinzhu/gorm/issues/1152.
The solution is to have Global Updates blocked by default from v2.
In the meantime you can set it explicitly: 
```go
db.BlockGlobalUpdate(true)
```

**Comment by sgon00** [comment]
When will v2 be released? I can not find any information about it.
